### PR TITLE
Ensure .sample() is detached and .rsample() requires grad

### PR DESCRIPTION
**Pull request by fritzo**
This fixes a bug in `TransformedDistribution` whereby `.sample()` was accidentally reparameterized. It also fixes a couple other minor `.sample()` bugs and adds to tests to distinguish the behavior of `.sample()` and `.rsample()`.

### CODE REVIEW

**Comment by fritzo** [comment]
@alicanb @rachit Does anyone want to review this minor bugfix PR before I send upstream?

**Comment by alicanb** [comment]
I'm taking a look right now

File: torch/distributions/poisson.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def __init__(self, rate):
 
     def sample(self, sample_shape=torch.Size()):
         shape = self._extended_shape(sample_shape)
-        return _poisson(self.rate.expand(shape))
+        return _poisson(self.rate.expand(shape)).detach()
```
Comment (alicanb) [inline torch/distributions/poisson.py]: I think all those `.detach()`s will raise error if parameters no distributions are not `Variable`

File: torch/distributions/transformed_distribution.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def sample(self, sample_shape=torch.Size()):
         x = self.base_dist.sample(sample_shape)
         for transform in self.transforms:
             x = transform(x)
-        return x
+        return x.detach() if hasattr(x, 'detach') else x
```
Comment (rachtsingh) [inline torch/distributions/transformed_distribution.py]: Is this equivalent to `isinstance(x, Variable)` but just future-proof to the Tensor/Variable merge?

Comment (fritzo) [reply torch/distributions/transformed_distribution.py]: I think it's equivalent, but this version is more literally exactly what I want. It is copied from `Distribution.sample()`.

**Comment by rachtsingh** [comment]
LGTM! Unfortunately that calls to `torch.bernoulli`, etc. don't automatically detach the output (since there's no gradient).

**Comment by alicanb** [comment]
> Is this equivalent to `isinstance(x, Variable)` but just future-proof to the Tensor/Variable merge?
I think this is to keep things consistent with `Distribution.sample()`, but I know @fritzo doesn't like `hasattr` too much, we can change them both if you like @fritzo ?

**Comment by fritzo** [comment]
> we can change them both

I'll just defer to the upstream reviewers who are informed on the State of Variables in PyTorch.

Thanks for reviewing!

**Comment by fritzo** [comment]
@alicanb may I move this upstream? Github says you've requested changes but it's not clear to me what you'd like changed.

File: torch/distributions/poisson.py (hunk 1 above)
Comment (alicanb) [reply torch/distributions/poisson.py]: @fritzo here

Comment (fritzo) [reply torch/distributions/poisson.py]: Ok, I'll use the `hasattr(x, 'detach')` trick more consistently. If upstream reviewers say we can assume everything is a variable, I'll revert this and also `Distribution.sample()`.
